File: source_files/haskell/Q1-10-lists/p08.hs
-- Eliminate consecutive duplicates of list elements. 
-- If a list contains repeated elements tehy shoudl be replaced with a single copy of the element. The order of elements should not be changed. 

-- With Data.List

import Data.List

compress :: Eq a => [a] -> [a]
compress = map head . group

-- With list comprehension. 
compress' :: Eq a => [a] -> [a]
compress' [] = []
compress' (x:[]) = [x]
compress' (a:b:xs)
    | a == b = compress' (b:xs)
    | otherwise = a : compress' (b:xs)

-- With fold. 
compress'' :: Eq a => [a] -> [a]
compress'' list = foldr (\a (b:bs) -> if a == b
                                    then (b:bs)
                                else (a:b:bs)) [last list] list

File: ben/prolog/P01-P28-lists/p08.pl
% Eliminate consecutive duplicates of list elements.

compress([], []).
compress([X], [X]).
compress([X, X|Xs], R) :- 
    compress([X|Xs], R).
compress([X, Y|Ys], [X|R]) :- 
    X \= Y,
    compress([Y|Ys], R).
